service/forward_service.cc

```diff
@@ -0,0 +1,210 @@
+#include "cql3/selection/selectable.hh"
+#include "cql3/selection/selectable-expr.hh"
+#include "cql3/selection/selection.hh"
+#include "cql3/selection/raw_selector.hh"
+#include "cql3/cql_config.hh"
+#include "cql3/column_identifier.hh"
+#include "cql3/query_options.hh"
+#include "cql3/result_set.hh"
+#include "cql_serialization_format.hh"
+
+namespace service {
+static constexpr int DEFAULT_INTERNAL_PAGING_SIZE = 10000;
+static logging::logger flogger("forward_service");
+
+static inline
+const dht::token& end_token(const dht::partition_range& r) {
+    static const dht::token max_token = dht::maximum_token();
+    return r.end() ? r.end()->value().token() : max_token;
+}
+
+static inet_address_vector_replica_set get_live_endpoints(keyspace& ks, const dht::token& token) {
+    auto erm = ks.get_effective_replication_map();
+    inet_address_vector_replica_set eps = erm->get_natural_endpoints_without_node_being_replaced(token);
+    auto itend = boost::range::remove_if(eps, std::not1(std::bind1st(std::mem_fn(&gms::gossiper::is_alive), &gms::get_local_gossiper())));
+    eps.erase(itend, eps.end());
+    return eps;
+}
+
+locator::token_metadata_ptr forward_service::get_token_metadata_ptr() const noexcept {
+    return _shared_token_metadata.get();
+}
+
+future<> forward_service::stop() {
+    return make_ready_future<>();
+}
+
+// Due to `cql3::selection::selection` not being serializable, it cannot be
+// stored in `forward_request`. It has to mocked on the receiving node,
+// based on requested reduction types.
+static shared_ptr<cql3::selection::selection> mock_selection(const std::vector<query::forward_request::reduction_type>& reduction_types, schema_ptr schema, database& db) {
+    std::vector<shared_ptr<cql3::selection::raw_selector>> raw_selectors;
+
+    auto mock_singular_selection = [&] (const query::forward_request::reduction_type& type) {
+        switch (type) {
+            case query::forward_request::reduction_type::count: {
+                auto selectable = cql3::selection::make_count_rows_function_expression();
+                auto column_identifier = make_shared<cql3::column_identifier>("count", false);
+                return make_shared<cql3::selection::raw_selector>(selectable, column_identifier);
+            }
+            default:
+                throw std::runtime_error("not implemented reduction type");
+        }
+    };
+
+    for (auto const& type : reduction_types) {
+        raw_selectors.emplace_back(mock_singular_selection(type));
+    }
+
+    return cql3::selection::selection::from_selectors(db, schema, std::move(raw_selectors));
+}
+
+future<query::forward_result> forward_service::execute(query::forward_request req) {
+    schema_ptr schema = local_schema_registry().get(req.cmd.schema_version);
+    auto command = make_lw_shared<query::read_command>(std::move(req.cmd));
+
+    flogger.debug("forward_request: received req={} from addr={}", req, _messaging.listen_address());
```

Comment:
Something important here is missing that I missed earlier - this is not parallel across shards. So if you have a 10 node cluster with 60 shards each, runtime will decrease by a factor of 10, not 600.

We also need to parallelize execution across shards.

Reply:
I don't think I follow... after a few discussions it was decided that splitting the range to single vnodes makes sense, so on the dispatching side (ref: `forward_service::dispatch`) the range will be split into single vnodes and performed in parallel. Now, when this function, `forward_service::execute`, is ran on the receiving side with a single vnode range, this node will simply perform a range read on a range that it owns, because we only send requests to vnode owners. Wouldn't the work be distributed to all shards locally in this case? I always imagined that it's how reading a range works on a replica (well, the receiving node is technically a coordinator, but it will attempt to read locally, assuming cl==ONE for simplicity).

The original idea was to split the range not by vnodes but by a-range-that-each-shard-owns, but it was discarded due to being likely to produce too many too small ranges, especially assuming that a node is capable of reading its own vnode in an efficient way.

Reply:
Each vnode will be processed sequentially on the shards.

Splitting by vnode is correct, but we should also parallelize the query per shard on each node. There are often more shards per node than nodes!

We don't have a way to say "read this token range but only take partitions for this shard", apart from splitting the token range by shard. I proposed an extension [here](https://groups.google.com/g/scylladb-dev/c/14Me4HpztxI/m/RhTRRwNJBwAJ).

Reply:
Ok, I get this idea of having parallelism of |vnodes|*|shard-count|, but I don't get why the split should happen on the node which dispatches the request (a.k.a. the supercoordinator).
If I understand the suggestion above correctly, it's supposed to work as follows:
1. Client requests an aggregation query on token range R
2. The range R is split into vnodes A, B, C
3. Each vnode is split into a set of tokens that a particular shard owns, so for 4 shards we end up with A', A'', A''', A'''', B', B'', B''', B'''', C', C'', C''', C''''
4. We send a request for each of this set in parallel to nodes that own them
5. We reduce the result

Is that correct? I'll continue this response assuming that it is, so please correct me if it's not the case.

Anyway, in this scenario we have 3 vnodes and 4 shards, which translates to sending 12 rpc requests. It sounds wasteful given that these requests share most of the information, and only differerence is the shard we're interested in. What makes the situation worse is that if the nodes don't have matching shard counts, we need to recalculate stuff so that it matches what the sender wanted. It sounds a little convoluted, especially that in this case what the coordinator ultimately wants is the results from the whole vnode.

> Each vnode will be processed sequentially on the shards.

I suppose my whole confusion revolves around this sentence. Why? Does it **need** to be processed sequentially? What we could do instead is:
1. Split range R into vnodes A, B, C
2. We send a request for each vnode, with a special annotation: "I don't care about the order at all, I'll be reducing partial results later anyway. Feel free to fetch this information in parallel on all shards locally and send everything back to me; I repeat, I don't care if it's processed sequentially"
3. We reduce the result

What we gain here is:
1. The dispatcher doesn't care what's the shard count of the target node - the target knows the number of its own shards, so it's better suited to distribute the work
2. We only send |vnodes| requests instead of |vnodes|*|shard-count-of-the-dispatcher|

Reply:
> Anyway, in this scenario we have 3 vnodes and 4 shards, which translates to sending 12 rpc requests. 

This actually ends up being 3 * 4096 because our shard ownership is decided according to the 12 lower bits of tokens.

Reply:
@denesb not according to the idea described above - true, it would translate to 3*4096 (or something similar) contiguous ranges, but a single request would ask not for a range, but for a set, consisting of all ranges from a particular vnode that belong to a shard, so at least the 4096 fanout is handled locally. My claim is that even the shard-count fanout should be pushed to replicas, especially that we then avoid the problem of nodes heterogeneous wrt. shard count.

Reply:
If you do something like that, you can further reduce the number of request, sending a single request per node, the node taking care of parallelizing vnodes and shards (probably only shard-level concurrency makes sense).

Reply:
It does not have to happen on the supercoordinator. A coordinator can just as easily perform a local invoke_on_all() per vnode.

btw, we could send |nodes| requests by sending a partition_range_vector to each node with all the vnodes it will be responsible for, rather than sending a request per vnode. This will reduce artificial priority inflation caused by running many requests in parallel on behalf of one query.

Reply:
Ugh, I repeated @denesb comment but with more words.

Reply:
@avikivity @denesb I think that splitting per vnode still has some merit. Provided RF > 1, we can more easily distribute requests to multiple coordinators, by sending not only to primary owners, but any owners; that would provide even more gain if the request has cl==ONE. Even with cl > ONE, it will distribute the coordinator work more evenly.

Sending |nodes| requests is a definitive gain, but as long as we can afford |vnodes|, maybe the additional granularity will still be worth it, given the point I described above? We could of course go even further and split each partition_range_vector into |nodes| and then send, but that sounds a little overcomplicated to code, given that we already have utility functions which produce vnodes.

Reply:
*note for people who read it via e-mail notifications instead of github: I just replaced |nodes| with |owners| in the last paragraph

Reply:
I don't see why it helps to send to non-primary-owners (btw, we shouldn't send to them if they're down). The primary owners should be well distributed, and sending to non-primary owners can only disturb the distribution.

That reminds me, we should look for local primary owners only if the replication strategy is data center aware and if the queries are local.

Reply:
Converting from vnodes to owners is just appending each owner's vnodes into a vector, and sorting.

Reply:
Ok, if the granularity does not help at all and can even hurt, then sending a single request per node sounds like the most reasonable path.

Reply:
@avikivity @denesb edit: I just spoke with @havaker and current code is actually almost there - it already groups vnodes per owner and sends a single dispatch request containing all the vnodes to each owner. So the only thing to implement from there is to optimize the dispatched request on the receiving side - paralellize wrt. local shard count.

Reply:
Ok. @havaker is lucky to have such a spokesperson!